### PR TITLE
Validate constructor results. Fixes issue #218.

### LINKED ISSUE #218: checkConstructorResult creates subtype constraints with null argument
See #217 for the version of CFI that produces the issue.

After pulling in typetools' CF and CFI master, opprop CFI will crash on test file as simple as
```
class I {}
```
because `org.checkerframework.common.basetype.BaseTypeVisitor.checkConstructorResult` will try to create a subtype constraint with null argument.

Sample stack trace:
```
  Exception: java.lang.Throwable; Stack trace: org.checkerframework.javacutil.BugInCF.<init>(BugInCF.java:14)
  checkers.inference.model.SubtypeConstraint.create(SubtypeConstraint.java:51)
  checkers.inference.model.ConstraintManager.createSubtypeConstraint(ConstraintManager.java:94)
  checkers.inference.model.ConstraintManager.addSubtypeConstraintNoErrorMsg(ConstraintManager.java:199)
  checkers.inference.InferenceQualifierHierarchy.isSubtype(InferenceQualifierHierarchy.java:225)
  org.checkerframework.common.basetype.BaseTypeVisitor.checkConstructorResult(BaseTypeVisitor.java:720)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitMethod(BaseTypeVisitor.java:587)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitMethod(BaseTypeVisitor.java:167)
  com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept(JCTree.java:800)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:68)
  org.checkerframework.framework.source.SourceVisitor.scan(SourceVisitor.java:88)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:301)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:167)
  com.sun.source.util.TreeScanner.scan(TreeScanner.java:91)
  com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:99)
  com.sun.source.util.TreeScanner.visitClass(TreeScanner.java:133)
  org.checkerframework.framework.source.SourceVisitor.visitClass(SourceVisitor.java:94)
  org.checkerframework.common.basetype.BaseTypeVisitor.processClassTree(BaseTypeVisitor.java:384)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:339)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:167)
  com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:720)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:50)
  org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:82)
  org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:1008)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:522)
  org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:182)
  com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:681)
  com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:111)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1342)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1296)
  com.sun.tools.javac.main.JavaCompiler.compile2(JavaCompiler.java:901)
  com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:860)
  com.sun.tools.javac.main.Main.compile(Main.java:523)
  com.sun.tools.javac.main.Main.compile(Main.java:381)
  com.sun.tools.javac.main.Main.compile(Main.java:370)
  com.sun.tools.javac.main.Main.compile(Main.java:361)
  checkers.inference.CheckerFrameworkUtil.invokeCheckerFramework(CheckerFrameworkUtil.java:12)
  checkers.inference.InferenceMain.startCheckerFramework(InferenceMain.java:189)
  checkers.inference.InferenceMain.run(InferenceMain.java:143)
  checkers.inference.InferenceMain.main(InferenceMain.java:115)
1 error
```

### PR DESCRIPTION
**Pull request by d367wang**
fixes #218

This PR works with downstream PR https://github.com/opprop/security-demo/pull/7 because security-demo is special in having non-top default.

### CODE REVIEW

File: src/checkers/inference/InferenceQualifierHierarchy.java [hunk 1]
```diff
@@ -330,25 +314,88 @@ public AnnotationMirror leastUpperBound(final AnnotationMirror a1, final Annotat
...
+            if (slot != null) {
+                annos.add(slotManager.getAnnotation(slot));
+            }
+        }
+
+        if (annos.size() != 1) {
+            throw new BugInCF(
+                    "There should be exactly 1 top qualifier in inference hierarchy"
+                            + "( checkers.inference.qual.VarAnnot ).\n"
+                            + "Tops found ( " + InferenceUtil.join(annos) + " )"
```
Comment (d367wang) [inline src/checkers/inference/InferenceQualifierHierarchy.java:364]: bottom

File: src/checkers/inference/InferenceQualifierHierarchy.java [hunk 2]
```diff
@@ -330,25 +314,88 @@ public AnnotationMirror leastUpperBound(final AnnotationMirror a1, final Annotat
...
+            );
+        }
+        return annos;
```
Comment (d367wang) [inline src/checkers/inference/InferenceQualifierHierarchy.java:367]: unmodified set

File: src/checkers/inference/InferenceQualifierHierarchy.java [hunk 3]
```diff
@@ -28,40 +28,24 @@
  */
 public class InferenceQualifierHierarchy extends MultiGraphQualifierHierarchy {
     private final InferenceMain inferenceMain = InferenceMain.getInstance();
-    private final AnnotationMirror varAnnot;
 
     private final SlotManager slotMgr;
     private final ConstraintManager constraintMgr;
 
+    /** Corresponding VarAnnot for the real top qualifier */
```
Comment (d367wang) [inline src/checkers/inference/InferenceQualifierHierarchy.java:35]: Unmodifiable singleton set containing the VarAnnot corresponding to the real top qualifier.

File: src/checkers/inference/InferenceQualifierHierarchy.java [hunk 4]
```diff
@@ -330,25 +314,88 @@ public AnnotationMirror leastUpperBound(final AnnotationMirror a1, final Annotat
...
+        return topVarAnnot;
+    }
+
+    @Override
+    public Set<? extends AnnotationMirror> getBottomAnnotations() {
+        return bottomVarAnnot;
+    }
+
     // ================================================================================
     // TODO Both of these are probably wrong for inference. We really want a new VarAnnot for that position.
```
Comment (d367wang) [inline src/checkers/inference/InferenceQualifierHierarchy.java:381]: remove

File: src/checkers/inference/solver/constraintgraph/GraphBuilder.java [hunk 5]
```diff
@@ -34,13 +34,9 @@
     private final ConstraintGraph graph;
     private AnnotationMirror top;
```
Comment (d367wang) [inline src/checkers/inference/solver/constraintgraph/GraphBuilder.java:35]: final

**Review by wmdietl — APPROVED** [review]
Thanks!

**Pull request merged**